Toolchain update (GCC 13.1.1, binutils 2.40)

vhotspur opened the pull request:
I was unable to build our current toolchain on Fedora 37 so I decided to try to update it to something newer. This PR contains all changes required to make it work with newer GCC and binutils.

I would appreciate if someone would review the changes and test them (I have not yet tested it across all architectures).

A lot of changes in the code are related to better warnings from the compiler: some actually revealed true bugs that I tried to fix, some has to be suppressed.

Note that `cpptest` fails miserably but it seems to me that the same is happening on `master` branch too. Not sure why, it looks to me like if the vtable gets corrupted somehow.

jxsvoboda commented:
Hi Vojta!

Thanks a lot for working on this!

Looks good to me overall. I added some small nits. I am not entirely sure I understand the changes to lexicographical_compare.

Thanks!
Jiri

le-jzr commented:
Thanks for working on this. Added some small comments but looks good to me!

le-jzr commented:
@jxsvoboda You mention nits but I don't see any other comments from you. Is that a problem with my app/eyes or did you not send any?

jermar commented:
Hi Vojta, I could build the toolchain for all architectures, but am gettin errors when building HelenOS/arm32/integratorcp:

[33/2007] Compiling C object kernel/libnoninstrumentables.a.p/arch_arm32_src_atomic.c.o
FAILED: kernel/libnoninstrumentables.a.p/arch_arm32_src_atomic.c.o 
/usr/local/cross/bin/arm-helenos-gcc -Ikernel/libnoninstrumentables.a.p -I../../../kernel/generic/include -I../../../kernel/genarch/include -I../../../kernel/arch/arm32/include -I../../../abi/arch/arm32/include -I../../../abi/include -Ikernel/test -I../../../kernel/test -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -std=gnu11 -imacros /home/jermar/software/HelenOS/helenos/build_all/arm32/integratorcp/config.h -O3 -fexec-charset=UTF-8 -finput-charset=UTF-8 -Wa,--fatal-warnings -Wall -Wextra -Wwrite-strings -Wunknown-pragmas -Wno-unused-parameter -pipe -ffunction-sections -fno-common -fdebug-prefix-map=/home/jermar/software/HelenOS/helenos=. -Wsystem-headers -fvar-tracking-assignments -Werror -Wmissing-prototypes -Werror-implicit-function-declaration -Wno-missing-braces -Wno-missing-field-initializers -Wno-unused-parameter -Wno-clobbered -Wno-nonnull-compare -fno-builtin-strftime -mcpu=arm926ej-s -D__LE__ -fno-omit-frame-pointer -mapcs-frame -ffixed-r9 -mtp=soft -mfloat-abi=hard -mno-unaligned-access -mfpu=vfpv3 -imacros /home/jermar/software/HelenOS/helenos/build_all/arm32/integratorcp/config.h -D_HELENOS_SOURCE -DKERNEL -DHELENOS_RELEASE=0.12.1 '-DHELENOS_COPYRIGHT=Copyright (c) 2001-2022 HelenOS project' -DHELENOS_CODENAME=Cathode -D__32_BITS__ -ffreestanding --param=min-pagesize=0 -Wno-cast-function-type -MD -MQ kernel/libnoninstrumentables.a.p/arch_arm32_src_atomic.c.o -MF kernel/libnoninstrumentables.a.p/arch_arm32_src_atomic.c.o.d -o kernel/libnoninstrumentables.a.p/arch_arm32_src_atomic.c.o -c ../../../kernel/arch/arm32/src/atomic.c
../../../kernel/arch/arm32/src/atomic.c:40:10: error: conflicting types for built-in function '__atomic_fetch_add_4'; expected 'unsigned int(volatile void *, unsigned int,  int)' [-Werror=builtin-declaration-mismatch]
   40 | unsigned __atomic_fetch_add_4(volatile unsigned *mem, unsigned val, int model)
      |          ^~~~~~~~~~~~~~~~~~~~
../../../kernel/arch/arm32/src/atomic.c:53:10: error: conflicting types for built-in function '__atomic_fetch_sub_4'; expected 'unsigned int(volatile void *, unsigned int,  int)' [-Werror=builtin-declaration-mismatch]
   53 | unsigned __atomic_fetch_sub_4(volatile unsigned *mem, unsigned val, int model)
      |          ^~~~~~~~~~~~~~~~~~~~
../../../kernel/arch/arm32/src/atomic.c:69:7: error: conflicting types for built-in function '__sync_val_compare_and_swap_4'; expected 'unsigned int(volatile void *, unsigned int,  unsigned int)' [-Werror=builtin-declaration-mismatch]
   69 | void *__sync_val_compare_and_swap_4(void **ptr, void *expected, void *new_val)
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../kernel/arch/arm32/src/atomic.c:98:7: error: conflicting types for built-in function '__atomic_compare_exchange_4'; expected '_Bool(volatile void *, void *, unsigned int,  _Bool,  int,  int)' [-Werror=builtin-declaration-mismatch]
   98 | _Bool __atomic_compare_exchange_4(void **mem, void **expected, void *desired, _Bool weak, int success, int failure)
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../kernel/arch/arm32/src/atomic.c:114:7: error: conflicting types for built-in function '__atomic_exchange_4'; expected 'unsigned int(volatile void *, unsigned int,  int)' [-Werror=builtin-declaration-mismatch]
  114 | void *__atomic_exchange_4(void **mem, void *val, int model)
      |       ^~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
[42/2007] Compiling C object kernel/libinstrumentables.a.p/generic_src_mm_as.c.o
ninja: build stopped: subcommand failed.

jermar commented:
Also, arm64/virt build ends with this error:
179/1495] Linking target kernel/kernel.1.elf
FAILED: kernel/kernel.1.elf 
/usr/local/cross/bin/aarch64-helenos-gcc  -o kernel/kernel.1.elf kernel/kernel.1.elf.p/meson-generated_.._empty_map.S.o -Wl,--as-needed -Wl,--no-undefined -Wl,--whole-archive -Wl,--start-group kernel/libinstrumentables.a kernel/libnoninstrumentables.a -Wl,--end-group -Wl,--no-whole-archive -Wl,--gc-sections -Wl,--warn-common -Wl,--fatal-warnings -O3 -fexec-charset=UTF-8 -finput-charset=UTF-8 -Wa,--fatal-warnings -Wall -Wextra -Wwrite-strings -Wunknown-pragmas -Wno-unused-parameter -pipe -ffunction-sections -fno-common -fdebug-prefix-map=/home/jermar/software/HelenOS/helenos=. -Wsystem-headers -fvar-tracking-assignments -Werror -Wmissing-prototypes -Werror-implicit-function-declaration -Wno-missing-braces -Wno-missing-field-initializers -Wno-unused-parameter -Wno-clobbered -Wno-nonnull-compare -fno-builtin-strftime -D__LE__ -fno-omit-frame-pointer -march=armv8-a+nofp+nosimd -mgeneral-regs-only -imacros /home/jermar/software/HelenOS/helenos/build_all/arm64/virt/config.h -D_HELENOS_SOURCE -DKERNEL -DHELENOS_RELEASE=0.12.1 '-DHELENOS_COPYRIGHT=Copyright (c) 2001-2022 HelenOS project' -DHELENOS_CODENAME=Cathode -D__64_BITS__ -ffreestanding --param=min-pagesize=0 -Wno-cast-function-type -nostdlib -Wl,--nmagic -Wl,--no-warn-rwx-segments -T /home/jermar/software/HelenOS/helenos/build_all/arm64/virt/kernel/_link.ld -Wl,-Map,/home/jermar/software/HelenOS/helenos/build_all/arm64/virt/kernel/kernel.1.elf.map
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libinstrumentables.a.p/generic_src_main_kinit.c.o: in function `kinit':
kinit.c:(.text.kinit+0x40c): undefined reference to `__aarch64_ldadd4_relax'
kinit.c:(.text.kinit+0x40c): relocation truncated to fit: R_AARCH64_CALL26 against undefined symbol `__aarch64_ldadd4_relax'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libinstrumentables.a.p/generic_src_mm_as.c.o: in function `as_release':
as.c:(.text.as_release+0x20): undefined reference to `__aarch64_ldadd4_rel'
as.c:(.text.as_release+0x20): relocation truncated to fit: R_AARCH64_CALL26 against undefined symbol `__aarch64_ldadd4_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libinstrumentables.a.p/generic_src_mm_as.c.o: in function `as_switch':
as.c:(.text.as_switch+0x13c): undefined reference to `__aarch64_ldadd4_relax'
as.c:(.text.as_switch+0x13c): relocation truncated to fit: R_AARCH64_CALL26 against undefined symbol `__aarch64_ldadd4_relax'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libinstrumentables.a.p/generic_src_mm_slab.c.o: in function `slab_alloc':
slab.c:(.text.slab_alloc+0x19c): undefined reference to `__aarch64_ldadd8_acq_rel'
slab.c:(.text.slab_alloc+0x19c): relocation truncated to fit: R_AARCH64_CALL26 against undefined symbol `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: slab.c:(.text.slab_alloc+0x220): undefined reference to `__aarch64_ldadd8_acq_rel'
slab.c:(.text.slab_alloc+0x220): relocation truncated to fit: R_AARCH64_CALL26 against undefined symbol `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: slab.c:(.text.slab_alloc+0x29c): undefined reference to `__aarch64_ldadd8_acq_rel'
slab.c:(.text.slab_alloc+0x29c): relocation truncated to fit: R_AARCH64_CALL26 against undefined symbol `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: slab.c:(.text.slab_alloc+0x38c): undefined reference to `__aarch64_ldadd8_acq_rel'
slab.c:(.text.slab_alloc+0x38c): relocation truncated to fit: R_AARCH64_CALL26 against undefined symbol `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libinstrumentables.a.p/generic_src_mm_slab.c.o: in function `magazine_obj_put':
slab.c:(.text.magazine_obj_put+0xe8): undefined reference to `__aarch64_ldadd8_acq_rel'
slab.c:(.text.magazine_obj_put+0xe8): relocation truncated to fit: R_AARCH64_CALL26 against undefined symbol `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libinstrumentables.a.p/generic_src_mm_slab.c.o:slab.c:(.text.magazine_obj_put+0x154): more undefined references to `__aarch64_ldadd8_acq_rel' follow
kernel/libinstrumentables.a.p/generic_src_mm_slab.c.o: in function `magazine_obj_put':
slab.c:(.text.magazine_obj_put+0x154): relocation truncated to fit: R_AARCH64_CALL26 against undefined symbol `__aarch64_ldadd8_acq_rel'
kernel/libinstrumentables.a.p/generic_src_mm_slab.c.o: in function `slab_obj_destroy':
slab.c:(.text.slab_obj_destroy+0x160): relocation truncated to fit: R_AARCH64_CALL26 against undefined symbol `__aarch64_ldadd8_acq_rel'
slab.c:(.text.slab_obj_destroy+0x1c0): additional relocation overflows omitted from the output
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_wait_internal':
thread.c:(.text.thread_wait_internal+0x54): undefined reference to `__aarch64_cas4_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_ready':
thread.c:(.text.thread_ready+0xbc): undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: thread.c:(.text.thread_ready+0xc8): undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_wait_timeout_callback':
thread.c:(.text.thread_wait_timeout_callback+0x1c): undefined reference to `__aarch64_swp4_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_create':
thread.c:(.text.thread_create+0x230): undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_put':
thread.c:(.text.thread_put+0x1c): undefined reference to `__aarch64_ldadd4_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_attach':
thread.c:(.text.thread_attach+0xbc): undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_exit':
thread.c:(.text.thread_exit+0x78): undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_interrupt':
thread.c:(.text.thread_interrupt+0x28): undefined reference to `__aarch64_swp4_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_wait_start':
thread.c:(.text.thread_wait_start+0x24): undefined reference to `__aarch64_swp4_acq'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_wakeup':
thread.c:(.text.thread_wakeup+0x1c): undefined reference to `__aarch64_swp4_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_try_get':
thread.c:(.text.thread_try_get+0x58): undefined reference to `__aarch64_cas4_relax'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_proc_thread.c.o: in function `thread_stack_trace':
thread.c:(.text.thread_stack_trace+0x48): undefined reference to `__aarch64_cas4_relax'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: thread.c:(.text.thread_stack_trace+0xcc): undefined reference to `__aarch64_swp4_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_synch_spinlock.c.o: in function `spinlock_lock':
spinlock.c:(.text.spinlock_lock+0x54): undefined reference to `__aarch64_swp1_acq'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: spinlock.c:(.text.spinlock_lock+0x88): undefined reference to `__aarch64_swp1_acq'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_synch_spinlock.c.o: in function `spinlock_unlock':
spinlock.c:(.text.spinlock_unlock+0x1c): undefined reference to `__aarch64_swp1_relax'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_synch_spinlock.c.o: in function `spinlock_trylock':
spinlock.c:(.text.spinlock_trylock+0x28): undefined reference to `__aarch64_swp1_acq'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_synch_spinlock.c.o: in function `spinlock_locked':
spinlock.c:(.text.spinlock_locked+0x18): undefined reference to `__aarch64_swp1_relax'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_console_cmd.c.o: in function `cmd_mcall0':
cmd.c:(.text.cmd_mcall0+0x80): undefined reference to `__aarch64_ldadd4_relax'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/generic_src_ipc_kbox.c.o: in function `ipc_connect_kbox':
kbox.c:(.text.ipc_connect_kbox+0x50): undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kbox.c:(.text.ipc_connect_kbox+0x74): undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kbox.c:(.text.ipc_connect_kbox+0x14c): undefined reference to `__aarch64_ldadd4_relax'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/test_atomic_atomic1.c.o: in function `test_atomic1':
atomic1.c:(.text.test_atomic1+0x38): undefined reference to `__aarch64_ldadd4_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: atomic1.c:(.text.test_atomic1+0x70): undefined reference to `__aarch64_ldadd4_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: atomic1.c:(.text.test_atomic1+0xa8): undefined reference to `__aarch64_ldadd4_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: atomic1.c:(.text.test_atomic1+0xe0): undefined reference to `__aarch64_ldadd4_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/test_mm_falloc2.c.o: in function `falloc':
falloc2.c:(.text.falloc+0x11c): undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: falloc2.c:(.text.falloc+0x13c): undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: falloc2.c:(.text.falloc+0x2b8): undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: falloc2.c:(.text.falloc+0x2c4): undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/local/cross/lib/gcc/aarch64-helenos/13.1.1/../../../../aarch64-helenos/bin/ld: kernel/libnoninstrumentables.a.p/test_synch_semaphore1.c.o: in function `producer':
semaphore1.c:(.text.producer+0x30): undefined reference to `__aarch64_ldadd8_acq_rel'

jermar commented:
On ia64 I get:
7/9] Linking target boot/image.boot.elf
FAILED: boot/image.boot.elf 
/usr/local/cross/bin/ia64-helenos-gcc  -o boot/image.boot.elf boot/components.o -Wl,--as-needed -Wl,--no-undefined -Wl,--whole-archive -Wl,--start-group boot/libbootfiles.a -Wl,--end-group -Wl,--no-whole-archive -Wl,--gc-sections -Wl,--warn-common -Wl,--fatal-warnings -O3 -fexec-charset=UTF-8 -finput-charset=UTF-8 -Wa,--fatal-warnings -Wall -Wextra -Wwrite-strings -Wunknown-pragmas -Wno-unused-parameter -pipe -ffunction-sections -fno-common -fdebug-prefix-map=/home/jermar/software/HelenOS/helenos=. -Wsystem-headers -Werror -Wmissing-prototypes -Werror-implicit-function-declaration -Wno-missing-braces -Wno-missing-field-initializers -Wno-unused-parameter -Wno-clobbered -Wno-nonnull-compare -fno-builtin-strftime -D__LE__ -fno-unwind-tables -mfixed-range=f32-f127 -mno-pic -mno-sdata -imacros /home/jermar/software/HelenOS/helenos/build_all/ia64/ski/config.h -D_HELENOS_SOURCE -DBOOT -DHELENOS_RELEASE=0.12.1 '-DHELENOS_COPYRIGHT=Copyright (c) 2001-2022 HelenOS project' -DHELENOS_CODENAME=Cathode -D__64_BITS__ -ffreestanding -nostdlib -Wl,--nmagic -T /home/jermar/software/HelenOS/helenos/build_all/ia64/ski/boot/_link.ld -Wl,--no-gc-sections -Wl,-Map,/home/jermar/software/HelenOS/helenos/build_all/ia64/ski/boot/image.boot.map
/usr/local/cross/lib/gcc/ia64-helenos/13.1.1/../../../../ia64-helenos/bin/ld: warning: boot/image.boot.elf has a LOAD segment with RWX permissions
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

jermar commented:
ppc32 fails too on:
/usr/local/cross/lib/gcc/ppc-helenos/13.1.1/../../../../ppc-helenos/bin/ld: warning: uspace/lib/libc.so.0.0 has a LOAD segment with RWX permission

riscv64 fails early on:
./../kernel/generic/src/interrupt/interrupt.c:83:35: error: array subscript n is outside array bounds of 'exc_table_t[0]' [-Werror=array-bounds=]
   83 |         iroutine_t old = exc_table[n].handler;

amd64, ia32, mips32 and sparc64 work fine and produce usable images.

le-jzr commented:
Hmm, libc shouldn't have any RWX segments, and those atomics shouldn't need emulation on aargh64 I think.

Gonna have a peek at that when I get home to see if I can spot the problem.

le-jzr commented:
Does anyone know if there's any downside to just using `gdb-multiarch` from distribution repos instead of building a custom GDB for every architecture?

le-jzr commented:
Just to share what I found so far, ppc32 ABI uses a weird system where `.plt` starts out empty and the dynamic linker needs to write code in it. There are GCC options `-msecure-plt`/`-mbss-plt` which toggle this behavior. The correct solution here is imo to make `-msecure-plt` default for the target.

There's also an RWX section warning for at least boot/arm32. Similar to kernel, this is just due to how the custom linker script is written amd can be avoided with a bit of tinkering there (the compiler flags works fine for now tho).

le-jzr commented:
So, making use of secure plt by default seems quite straightforward. For old GCC 8.3 the following change in `config.gcc` was enough.
```
powerpc-*-helenos*)
	tm_file="rs6000/secureplt.h ${tm_file} dbxelf.h elfos.h rs6000/sysv4.h"
	...
```

It also needs a change in ppc32 relocation processing. Simply making `R_PPC_JMP_SLOT` act the same as `R_PPC_ADDR32` and removing the code generation seems to work correctly. There's the option to detect which version is used by the ELF file and support both variants, but I don't really see the benefit of keeping all that code for binaries nobody should be making anyway. Gonna open a separate pull request soon.

vhotspur commented:
Hi, again. Sorry for the radio silence. I am back with the rest of the changes.

I was working on a different machine and then I forgot about some issues when reenabling all the warnings, hence I missed the other architectures. Should be (almost) okay now.

>  I am not entirely sure I understand the changes to lexicographical_compare.

I am not very happy with [this commit fcbf101cd868](https://github.com/HelenOS/helenos/pull/226/commits/fcbf101cd868ee6da4b68f5c9c4ac496316e4d0e) but I was unable to make the template code compile hence I copied the code from another function and when the comparison `<` is used like that, the template works again. Someone more knowledgeable in C++ might be able to revert the code back to remove the code duplication.

> [...] but am gettin errors when building HelenOS/arm32/integratorcp [and others too] [...]

Fixed or workaround-ed.

Atomics: I have updated the function signatures to match the builtins in new GCC.

genisoimage: not related to the toolchain update but IMHO it might be useful for other too. BTW, shouldn't we have xorriso as the first one as that one seems to me to be most recently developed (both `genisoimage` and `mkisofs` looks like without any further development)?

On ARM32 I switched to softfp because it seems that is the choice in `libgcc.a` that we link against. Workaround but I do not know anything better.

Workarounds for RWX segments: Jirka is probably right about his suggestions but I do not know how to properly apply them.

If you have any improvements, please, feel free to push commits to this branch directly.

vhotspur commented on meson/part/compiler_args/meson.build:
```diff
@@ -109,6 +109,7 @@ endif
 
 extra_cppflags = extra_common_flags + [
 	'-fno-exceptions',
+	'-Wno-misleading-indentation',
```
Because I was lazy to edit all the C++ code to add the missing braces around single command blocks ;-).

But seriously: we should probably update the code to never use `if (..) x()` but only  `if (..) { x() }`. But I think that can be done later in another PR.

vhotspur commented on uspace/srv/hid/display/test/window.c:
```diff
@@ -1108,13 +1108,18 @@ PCUT_TEST(window_set_cursor)
+	// Check that invalid cursors cannot be set: ignore enum conversions
+	// as we are out-of-bounds
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wenum-conversion"
```
I believe there are more cases when we are using `enum` instead of macros. Perhaps in another PR?

BTW, it seems to me that there was one another comment from you but I cannot find it at the moment?

jermar commented:
Hi Vojta, here's another batch of results from me:

- I didn't rebuild the toolchains themselves as I believed there had been no changes since my last test round
- This time I was able to build all HelenOS architectures
- I DIDN'T test booting the following ones, so we don't know if they work: arm32/beagleboardxm, arm32/beaglebone, arm32/gta02, ia64/i460GX, riscv64
- From the rest the following ones have a confirmed good bootable image: amd64, arm32/raspberrypi, arm64/virt, mips32/malta-be, mips32/malta-le, mips32/msim, ppc32, sparc64/ultra, sparc64/niagara

- The rest have some issues:

- arm32/integratorcp hangs on:
Inflating components ... 
 0x80a00000|0x00a00000: kernel.elf.gz image (571908/168812 bytes)

- ia64/ski dies on an exception:
[init:ext4fs(8)] ext4fs: Accepting connections
Task init:vfs (6) killed due to an exception at program counter 0x40000000000260c0.
ar.bsp=0xe00000000416c148       ar.bspstore=0x60000000002f4058
ar.rnat=0x0     ar.rsc=0xc
ar.ifs=0x8000000000000288       ar.pfs=0xc000000000000288
cr.isr=0x400000000      cr.ipsr=0x1013080a6010
cr.iip=0x40000000000260c0, #0   (?)
cr.iipa=0x4000000000028160      (?)
cr.ifa=0x400    ()
Kill message: Page fault: 0x0000000000000400.

- I will try to test on arm32/beaglebone, arm32/gta02 and report back. I will also try to have a look at the ia64 exception.

jermar commented:
More test results:
- arm32/beaglebone hangs similar to integratoricp on:
Inflating components ...
 0x80a00000|0x80a00000: kernel.elf.gz image (505268/145137 bytes)

- arm32/gta02 works (to the extent it worked before)
- arm32/raspberrypi real hardware also works (before tested only the QEMU model)

jermar commented:
The hangs on arm32 integratorcp and beaglebone seem to correlate with CONFIG_FPU and the change this PR does to it (hard-float -> soft-float).

jermar commented on meson/arch/arm32/meson.build:
```diff
@@ -44,7 +44,8 @@ arch_uspace_c_args = [
 	# Use vfp32 to allow context save/restore of d16-d31 regs.
-	arch_uspace_c_args += [ '-mfloat-abi=hard' ]
+	# Need to use softfp because our libgcc.a uses it too
+	arch_uspace_c_args += [ '-mfloat-abi=softfp' ]
```
I suspect this is causing the hangs on integratorcp and beaglebone (and likely also beagleboard)

vhotspur replied:
Hi, Jakub, thanks for testing! I tried recompiling (from clean state) with `hard` but I am unable to get integratorcp to boot either. It worked for you?

jermar replied:
Unfortunately it doesn't work for me either. Strangely though the the only two boards that are failing are the ones that are configured to use FPU via CONFIG_FPU.

jermar replied:
I looked into this a little more and I think the problem is that boot now uses an FPU instruction while the FPU is disabled, which results in an Undefined exception and some erratic behavior which restarts loader in a dirty way which results in executing halt() by chance.

Trace 0: 0x7fe2b406b1c0 [00000400/00003e64/00000020/ff000201] inflate_dynamic
R00=00000013 R01=00000006 R02=0000000f R03=0000011e
R04=0000c0a4 R05=0000be00 R06=0000bb28 R07=0000d3e4
R08=0000d3e4 R09=00000000 R10=0000001e R11=0000c09c
R12=00000027 R13=0000bad0 R14=0000000a R15=00003e64
PSR=600001db -ZC- A und32
Taking exception 1 [Undefined Instruction] on CPU 0
...from EL1 to EL1
...with ESR 0x0/0x2000000
Trace 0: 0x7fe2b4000280 [00000400/00000004/00000020/ff000201] 

where in inflate_dynamic there is:

    3e64:       eddf0beb        vldr    d16, [pc, #940] @ 4218 <inflate_dynamic+0x770>

I think we need to prevent boot from being built with FPU instructions.

jermar replied:
Changing -mfloat-abi to soft (but not softfp) makes the problem non-reproducible. The compiler is probably hesitant to make the same optimization via a call to the soft-float library. But I still think that we should look for a way to completely prevent the compiler from generating FPU instructions for loader and the kernel. As for the kernel, it's not clear to me why we have -mfpu=vfpv3 in its arch_kernel_c_args, but that's an unrelated topic.

jermar commented on uspace/drv/nic/ar9271/wmi.c:
```diff
@@ -229,6 +229,10 @@ errno_t wmi_send_command(htc_device_t *htc_device, wmi_command_t command_id,
 	size_t buffer_size = header_size + command_length;
 	void *buffer = malloc(buffer_size);
+	if (buffer == NULL) {
+		usb_log_error("Failed to allocated WMI message buffer (out of memory).\n");
```
allocate

vhotspur replied:
Thanks. Fixed.

le-jzr commented:
I believe there's a way for libgcc to be built in multiple versions and GCC to automatically select the appropriate one.

Unfortunately other things are eating my time until the end of the month, but after that I'll be free to look into it again.

vhotspur commented:
Jakub, thanks again! I pushed some more changes, mainly to allow testing with current toolchain (GCC 8) on top of my modifications plus some updates to the FPU configuration.

At the moment, I can build everything and boot of integratorcp and Raspberry profiles in QEMU works (on Raspberry it is terminal only in kernel console but I guess that is okay). I have no way to test any of the arm32 builds on real hardware, though.

But I admit that it is becoming more and more a mumbo jumbo of "let's try that and see what it does" for me :-).

Jirka, that would be great. Thank you.

jermar commented:
Hi Vojta, I will be able to retest on the real again hw after Aug 14.

vhotspur commented:
I have changed the `arm32` configuration to use `softfp` instead.

I have tested (almost) all available emulators from `tools/ew.py` and:

* `arm64/virt` fails with `Synchronous Exception at 0x000000007C405FB8`
* `arm32/raspberrypi` boots in kernel console only (cannot switch to user terminal)
* `ia64/ski` page-faults in VFS, the counter points into some futex code
* `sparc64/niagara` not tested
* `riscv64` hangs on `Booting the kernel...`

Others look healthy.

jxsvoboda commented:
I successfully tested on a real BeagleBoard xM. Looking good! See my report on helenos-devel for details.

jermar commented:
Cannot reproduce the arm64/virt fail. Seems to work fine for me. arm32/raspberrypi behavior is expected. I can reproduce ia64/ski crash, but so far have no idea what is causing it. I had tested sparc64/niagara before and it worked fine with these changes. riscv64 I have no idea.

martin-decky commented:
The riscv64 port is still unfinished, thus there's nothing beyond `Booting the kernel...` and that should not be considered a blocker for this PR.

vhotspur commented:
> I successfully tested on a real BeagleBoard xM. Looking good! See my report on helenos-devel for details.

Thank you!

> Cannot reproduce the arm64/virt fail. Seems to work fine for me. 

Good, thanks!

> The riscv64 port is still unfinished, thus there's nothing beyond Booting the kernel... and that should not be considered a blocker for this PR.

OK, makes sense.

Thanks a lot to everyone for the testing!

So I assume that I can merge this and create ticket for `ia64/ski` failure as the only regression?

jxsvoboda commented:
Sounds good.

jxsvoboda commented:
I'd really like to thank you for doing this. I have an additional request, though. Since you've just gone through the process of porting our custom GCC and binutils targets to new version in HelenOS/gcc and Helenos/binutils-gdb, would you be able to document that process in HelenOS wiki? It probably requires some voodoo with git branches and it would be really nice if people didn't have to reinvent the wheel each time we need to do this...

vhotspur commented:
About the toolchain upgrading.

---

The process is very similar for both GCC and binutils-gdb projects.

1. Add another Git remote with upstream development.
   * `git://sourceware.org/git/binutils-gdb.git`
   * `git://gcc.gnu.org/git/gcc.git`
2. Find a branch on the upstream with the right version.
   * `binutils-2_40-branch`
   * `releases/gcc-13`
3. Create a new branch starting on the last commit from branch from the previous step.
   * `binutils-2_40-helenos`
   * `13_1_1-helenos`
4. Rebase changes in previous `*-helenos` branch onto the branch from the previous step.
   * There will be conflicts: mostly in various `*config*` files as new targets are added or obsoleted ones are removed. Usually resolving them is straightforward.
5. Update the version numbers in `toolchain.sh`.

Pro-tip :-)

* Set up the repositories in `toolchain.sh` to point to your local clones to speed-up the cloning.
  * Or setup the clones in the `downloads` directory and comment-out all Git operations.

---

Not sure if I remembered all the big steps but this is at least rough outline. If you believe even this little is useful, I will copy it to the Wiki.

jermar reviewed (approved):
Thanks Vojta! I think we are good to merge this. Possible issue shall be addressed lazily.

le-jzr commented:
Sorry about the long radio silence. I had a good long look at the GCC/binutils ports and everything looks good to me on that front. There's a couple things in here that I want to take another look tomorrow, like the packed structs that seem misused to me. A good chunk of commits here can (and should) be applied to master regardless of toolchain. The bug fixes and such. Mind if I cherry-pick and push those while taking a look at the rest?

le-jzr commented:
Merged most of it with little changes, the notable ones being that I folded together some of the revisions of prior commits, and I removed the offending packed attribute rather than layering another on top (https://github.com/HelenOS/helenos/commit/590cb6d21c857534d940c4c0a165a9997826a0b0).

That leaves two explicit casts between enums that I feel like should have a more elegant solution, which I'm trying to figure out.
And https://github.com/HelenOS/helenos/pull/226/commits/74332347fd59477d87b1f568a103474efdf55696 which I don't understand the reason for. The commit message says it's "unused", but the code is very much used, so if it's unused in some other way I'd like a better explanation of the reason.

le-jzr commented:
Did a bit of cleanup in the toolchain ports, and created new updated branches with binutils 2.41 and GCC 13.2.

martin-decky commented:
BTW: Since GDB supports multi-arch now and most distributions actually distribute a multi-arch build of GDB, would it make sense to remove GDB from our toolchain script and thus save some time and disk space? Or does anyone see a major benefit of keeping our own GDB builds?

I vaguely remember discussing this already somewhere, but I am unable to find a written record of it. Thus let's reiterate here. Maybe we can build the GDB optionally.

le-jzr commented:
Yeah, I don't really see the point of building custom GDB. I've been using packaged multiarch whenever I needed it and haven't seen any issue. Although admittedly my use of GDB is rare and rather surface-level.

BTW, since we're talking about debuggers, I'm currently working on using DWARF data for stack traces. I also plan to look at making a rudimentary integrated debugger for the kernel, using the same. Basic stuff like inspecting stack frames and printing out variables after a panic.

martin-decky commented:
The DWARF support and the debugger plans are great news! Looking forward to it. And to the removal of the fragile map file parsing :)

If we only need symbol information, we might also use CTF which is supposed to be simpler to parse than DWARF at run time.

le-jzr commented:
> If we only need symbol information, we might also use CTF which is supposed to be simpler to parse than DWARF at run time.

Well, for just symbol names, we could also use the raw `.symtab` and `.strtab` sections. But I do like the idea of having proper filename:line:column output, so I'm aiming for that. :)

le-jzr commented:
> And [7433234](https://github.com/HelenOS/helenos/commit/74332347fd59477d87b1f568a103474efdf55696) which I don't understand the reason for. The commit message says it's "unused", but the code is very much used, so if it's unused in some other way I'd like a better explanation of the reason.

Ooooh, I get it now. It's just broken syntax and never affected the compilation. I guess that explains my past inability to get libatomic and libstdc++ to build. :facepalm:

le-jzr commented:
Done. Updated toolchain.sh in master to binutils 2.41, GDB 13.2 and GCC 13.2. Everything seems to build fine on my end, and don't see any issues running emulators so far, so hopefully there's no massive breakage in that extra minor version.
Whoever has access to ci.helenos.org should update it, including the old Meson there.

I was hoping that after the merge commit above, github will recognize the pull request as merged. Seems it's not that clever, so the hard way it is.

Thanks @vhotspur for all the work you did!
